src/librustc/ty/instance.rs

```diff
@@ -41,8 +42,12 @@ pub enum InstanceDef<'tcx> {
     /// drop_in_place::<T>; None for empty drop glue.
     DropGlue(DefId, Option<Ty<'tcx>>),
 
-    ///`<T as Clone>::clone` shim.
-    CloneShim(DefId, Ty<'tcx>),
+    ///`<T as Clone>::clone` shim for Copy types
+    CloneCopyShim(DefId, Ty<'tcx>),
+    ///`<T as Clone>::clone` shim for arrays and tuples
+    CloneStructuralShim(DefId, Ty<'tcx>),
+    ///`<T as Clone>::clone` shim for closures
+    CloneNominalShim(DefId, Ty<'tcx>),
```

Comment:
Why is this not `CloneClosureShim`?

Reply:
And (in all cases) can you comment what the `DefId` refers to? (If you're feeling heroic, do the same for the other variants)

Reply:
It used to be CloneClosureShim and eddyb had me change it to use a more generic name 😄

Reply:
It's literally `Clone::clone` - all of the variants (have to) record the original `DefId` that was passed to `Instance::resolve` *or* something more specific (mostly for `InstanceDef::Item`.